### PR TITLE
Add typo tolerance customization settings

### LINKED ISSUE #313: Add support to the typo tolerance customization
⚠️ This issue is generated, it means the examples and the namings do not necessarily correspond to the language of this repository.
Also, if you are a maintainer, feel free to add any clarification and instruction about this issue.

Related to:
- parent issue: https://github.com/meilisearch/integration-guides/issues/190
- specification: https://github.com/meilisearch/specifications/pull/117

----

Add a `typoTolerance` index settings resource to manage customization of the typo tolerance feature at the index level.

- Expose `typoTolerance` object on `/indexes/:index_uid/settings` endpoints.
- Add `GET/POST/DELETE` - `/indexes/:index_uid/settings/typo-tolerance` endpoints.

The `typoTolerance` object definition is made of the following properties:
  - `enabled` type (`boolean`)
  - `disableOnAttributes` type (`array[string]`)
  - `disableOnWords` type (`array[string]`)
  - `minWordSizeForTypos` type (`object`)
    - `oneTypo` type (`int`)
    - `twoTypos` type (`int`)

More information about what each field does could be [checked in the spec.](https://github.com/meilisearch/specifications/pull/117)

Check the other customization methods for reference like: `synonyms`, `searchableAttributes`, `filterableAttributes`.

TODO:
- [ ] Add the new methods to customize `typo-tolerance`
  - [ ] GET `/indexes/:index_uid/settings/typo-tolerance`
  - [ ] POST `/indexes/:index_uid/settings/typo-tolerance`
  - [ ] DELETE `/indexes/:index_uid/settings/typo-tolerance`
- [ ] Add tests

### PR DESCRIPTION
**Pull request by voloyev**
# Pull Request

## What does this PR do?
Fixes #313 

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: spec/meilisearch/index/settings_spec.rb [hunk 1]
```diff
@@ -722,4 +722,65 @@ def update_synonyms(index, synonyms)
...
+    end
+
+    before { client.create_index!(uid) }
+
+    it 'get default typo tolerance settings' do
+      task = index.typo_tolerance
+      client.wait_for_task(task['taskUid'])
+
+      expect(index.typo_tolerance).to eq(default_typo_tolerance)
+    end
```
Comment (brunoocasali) [inline spec/meilisearch/index/settings_spec.rb:761]: The `/typo-tolerance` sub-route does not respond with a task :)

```suggestion
    it 'gets default typo tolerance settings' do
      settings = index.typo_tolerance

      expect(settings).to eq(default_typo_tolerance)
    end
```

File: spec/meilisearch/index/settings_spec.rb [hunk 2]
```diff
@@ -722,4 +722,65 @@ def update_synonyms(index, synonyms)
...
+      client.wait_for_task(update_task['taskUid'])
+
+      task = index.typo_tolerance
+      client.wait_for_task(task['taskUid'])
+
+      reset_task = index.reset_typo_tolerance
+      client.wait_for_task(reset_task['taskUid'])
+
+      expect(index.typo_tolerance).to eq(default_typo_tolerance)
+    end
```
Comment (brunoocasali) [inline spec/meilisearch/index/settings_spec.rb:784]:
```suggestion
    it 'resets typo tolerance settings' do
      update_task = index.update_typo_tolerance(new_typo_tolerance)
      client.wait_for_task(update_task['taskUid'])

      reset_task = index.reset_typo_tolerance
      client.wait_for_task(reset_task['taskUid'])

      expect(index.typo_tolerance).to eq(default_typo_tolerance)
    end
```

File: spec/meilisearch/index/settings_spec.rb [hunk 3]
```diff
@@ -722,4 +722,65 @@ def update_synonyms(index, synonyms)
...
+
+    it 'update typo tolerance settings' do
+      update_task = index.update_typo_tolerance(new_typo_tolerance)
+      client.wait_for_task(update_task['taskUid'])
+
+      task = index.typo_tolerance
+      client.wait_for_task(task['taskUid'])
+
+      expect(index.typo_tolerance).to eq(new_typo_tolerance)
+    end
```
Comment (brunoocasali) [inline spec/meilisearch/index/settings_spec.rb:771]:
```suggestion
    it 'updates typo tolerance settings' do
      update_task = index.update_typo_tolerance(new_typo_tolerance)
      client.wait_for_task(update_task['taskUid'])

      expect(index.typo_tolerance).to eq(new_typo_tolerance)
    end
```

File: lib/meilisearch/index.rb [hunk 4]
```diff
@@ -393,5 +393,19 @@ def update_pagination(pagination)
...
       http_delete "/indexes/#{@uid}/settings/pagination"
     end
+
+    def typo_tolerance
+      http_get("/indexes/#{@uid}/settings/typo-tolerance")
+    end
+    alias get_typo_tolerance typo_tolerance
+
+    def update_typo_tolerance(typo_tolerance_attributes)
+      http_patch("/indexes/#{@uid}/settings/typo-tolerance", typo_tolerance_attributes)
```
Comment (brunoocasali) [inline lib/meilisearch/index.rb:403]: A good addition here is to use the utility method to transform everything in camel case, so we can use snake_case in our ruby/rails apps :)

It should look like this:

```rb
    def update_typo_tolerance(typo_tolerance_attributes)
      attributes = Utils.transform_attributes(typo_tolerance_attributes)

      http_patch("/indexes/#{@uid}/settings/typo-tolerance", attributes)
    end
```

**Review by brunoocasali — CHANGES_REQUESTED** [review]
Hello @voloyev, thanks a lot for your contribution!

I would like you to add some small things to make this PR perfect, do you mind?

Thanks for contributing with Meilisearch 👨‍🎤

**Comment by voloyev** [comment]
> Hello @voloyev, thanks a lot for your contribution!
> 
> 
> 
> I would like you to add some small things to make this PR perfect, do you mind?
> 
> 
> 
> Thanks for contributing with Meilisearch 👨‍🎤 

Hello! I'm do not mind☺️

**Comment by voloyev** [comment]
@brunoocasali Do I need to squash the commits?

**Comment by brunoocasali** [comment]
> @brunoocasali Do I need to squash the commits?

You can do it if you want, (I prefer to be honest), but it is not a requirement :)

**Comment by voloyev** [comment]
> > @brunoocasali Do I need to squash the commits?
> 
> 
> 
> You can do it if you want, (I prefer to be honest), but it is not a requirement :) 

Will do then later today

**Comment by voloyev** [comment]
> > > @brunoocasali Do I need to squash the commits?
> > 
> > 
> > You can do it if you want, (I prefer to be honest), but it is not a requirement :)
> 
> Will do then later today

Updated. Pls review

**Review by brunoocasali — APPROVED** [review]
Thank you for contributing with Meilisearch ❤️

**Comment by brunoocasali** [comment]
bors merge

**Pull request merged**